pkg/controller/egress_router/egress_router_controller.go

```diff
@@ -0,0 +1,173 @@
+	mgr    manager.Manager
+	client client.Client
+	status *statusmanager.StatusManager
+
+	egressrouters    map[types.NamespacedName]*egressrouter
+	egressrouterErrs map[types.NamespacedName]error
+}
+
+var ResyncPeriod = 5 * time.Minute
+
+func newEgressRouterReconciler(mgr manager.Manager, status *statusmanager.StatusManager) (reconcile.Reconciler, error) {
+
+	return &EgressRouterReconciler{
+		mgr:    mgr,
+		status: status,
+		client: mgr.GetClient(),
+
+		egressrouters:    map[types.NamespacedName]*egressrouter{},
+		egressrouterErrs: map[types.NamespacedName]error{},
+	}, nil
+}
+
+func (r EgressRouterReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log.Printf("Reconciling egressrouter.network.operator.openshift.io %s\n", request.NamespacedName)
+
+	obj := &netopv1.EgressRouter{}
+	err := r.mgr.GetClient().Get(context.TODO(), request.NamespacedName, obj)
+
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Printf("Egress Router %s seems to have been deleted\n", request.NamespacedName)
+			return reconcile.Result{}, nil
+		}
+		log.Println(err)
+		return reconcile.Result{}, err
+	}
+
+	// Check to see if we already know this object
+	existing := r.egressrouters[request.NamespacedName]
+	if existing != nil {
+		if !reflect.DeepEqual(obj.Spec, existing.spec) {
+			log.Printf("Egress Router %s has changed, refreshing\n", request.NamespacedName)
+			delete(r.egressrouters, request.NamespacedName)
+			existing = nil
+		}
+	}
+
+	if existing == nil {
+		log.Printf("Creating a new Egress Router")
+		objs, err := ensureEgressRouter(manifestDir)
+		for _, obj := range objs {
+			log.Printf("Applying manifest")
+			apply.ApplyObject(context.TODO(), r.client, obj)
+		}
+
+		if err != nil {
+			log.Println(err)
+			r.egressrouterErrs[request.NamespacedName] =
+				errors.Wrapf(err, "could not reconcile Egress Router %s", request.NamespacedName)
+			r.setStatus()
+			return reconcile.Result{}, err
+		}
+		r.egressrouters[request.NamespacedName] = existing
+	}
+	err = existing.sync()
+	if err != nil {
+		log.Println(err)
+		r.egressrouterErrs[request.NamespacedName] =
+			errors.Wrapf(err, "could not reconcile Egress Router %s", request.NamespacedName)
+		r.setStatus()
+		return reconcile.Result{}, err
+	}
+
+	log.Println("successful reconciliation")
+	delete(r.egressrouterErrs, request.NamespacedName)
+	r.setStatus()
+	return reconcile.Result{RequeueAfter: ResyncPeriod}, nil
+}
+
+// setStatus summarizes the status of all Egress Router objects and updates the statusmanager
+// as appropriate.
+func (r *EgressRouterReconciler) setStatus() {
+	if len(r.egressrouterErrs) == 0 {
+		r.status.SetNotDegraded(statusmanager.EgressRouterConfig)
+	} else {
+		msgs := []string{}
+		for _, e := range r.egressrouterErrs {
+			msgs = append(msgs, e.Error())
+		}
+		r.status.SetDegraded(statusmanager.EgressRouterConfig, "EgressRouterError", strings.Join(msgs, ", "))
+	}
+}
+
+type egressrouter struct {
```

Comment:
You never create any instances of this type. What does it do?

Reply:
it gets called in EgressRouterReconciler struct, pretty much as PKI does, what would you suggest in this case?